### PR TITLE
Improvement of the method `getCurrentScriptUID`.

### PR DESCRIPTION
**Pull request by leogedler**
- Added logic to the method `getCurrentScriptUID` to generate the UID hashing the script's src url and attributes if exist, if not, continue using a random UID.

- Added unit test for getCurrentScriptUID.

### CODE REVIEW

File: test/tests/dom/getCurrentScriptUID.js [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+    it('should create a data-uid-auto attribute', () => {
+        
+        getCurrentScriptUID();
+
+        const currentScript = getCurrentScript();
+        const uidAttr = currentScript.getAttribute(`${ ATTRIBUTES.UID }-auto`);
+
+        if (!uidAttr) {
+            throw new Error(
+                `Expected result have a 'data-uid-auto' attribute, got none`
```
Comment (gabrielo91) [inline test/tests/dom/getCurrentScriptUID.js:29]: Just a suggestion, maybe:

`should have returned 'data-uid-auto'`, got undefined

File: src/constants.js [hunk 2]
```diff
@@ -6,5 +6,10 @@ export const KEY_CODES = {
 };
 
 export const ATTRIBUTES = {
-    UID: 'data-uid'
+    UID:     'data-uid',
+    NONCE:   'csp-nonce',
```
Comment (gregjopa) [inline src/constants.js:10]:
```suggestion
    NONCE:   'data-csp-nonce',
```

Nonce is also prefixed with `data-` https://developer.paypal.com/docs/checkout/troubleshoot/support/#content-security-policy-rules

Comment (leogedler) [reply src/constants.js]: Got it, changed.

Probably we should update the docs as well? This table is a bit misleading:

image


https://developer.paypal.com/docs/checkout/reference/customize-sdk/#csp-nonce

File: src/dom.js [hunk 3]
```diff
@@ -1215,7 +1215,28 @@ export const getCurrentScriptUID : GetCurrentScriptUID = memoize(() => {
         return uid;
     }
 
-    uid = uniqueID();
+    if (script.src) {
+
+        let stringToHash = script.src;
+
+        Object.keys(ATTRIBUTES).forEach(key => {
```
Comment (gregjopa) [inline src/dom.js:1222]: @leogedler @bluepnume what do you think about ditching the data attribute allowed list and having the hash be `script.src` + `script.dataset` for all data attributes except uid? Ex:

Here's a script with a couple data attributes. Note how `dataset` converts the `data-` keys into camelCase and drops the prefix.
Screen Shot 2021-05-11 at 9 04 29 AM

```js
let stringToHash = script.src;

Object.keys(script.dataset).forEach(key => {
    // ignore uid attribute
    if (key === "uid") {
        return;
    }

    // every other data attribute is added to the stringHash
    stringToHash += script.dataset[key];
});
```

File: src/dom.js [hunk 4]
```diff
@@ -1215,7 +1215,24 @@ export const getCurrentScriptUID : GetCurrentScriptUID = memoize(() => {
...
-    uid = uniqueID();
+    if (script.src) {
+
+        let stringToHash = script.src;
+
+        Object.keys(script.dataset).forEach(key => {
+            if (key === 'uid') {
+                return;
+            }
+            stringToHash += script.dataset[key];
```
Comment (bluepnume) [inline src/dom.js:1226]: I think we should include the key name also. Can we do a `JSON.stringify({ url, attributes })` where attributes is a mapping of key->value?

File: src/dom.js [hunk 5]
```diff
@@ -1215,7 +1215,28 @@ export const getCurrentScriptUID : GetCurrentScriptUID = memoize(() => {
...
+            const attr = script.getAttribute(ATTRIBUTES[key]);
+            
+            if (attr) {
+                stringToHash += `${ attr }`;
+            }
+        });
+
+        const hashedString = hashStr(stringToHash);
+
+        uid = `UID_${ hashedString }`;
```
[the remark this answers is not in the document]
Comment (bluepnume) [reply src/dom.js:1231]: sounds good to me but let's add naming convention for all cases:

- `uid_hash_xyz`
- `uid_random_xyz`

[the remark this answers is not in the document]
Comment (gregjopa) [reply src/dom.js:1231]: Good suggestion @bluepnume! Let's implement it.

File: src/dom.js (hunk 4 above)
Comment (gregjopa) [reply src/dom.js:1226]: Whoops I forgot about the attribute keys. Good suggestion.

Comment (leogedler) [reply src/dom.js]: I have implemented the suggested approach.

File: src/dom.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (leogedler) [reply src/dom.js]: Got it, I have updated the logic to add the prefix for both cases, the hashing and the random uid.

**Review by gregjopa — APPROVED** [review]
Looks great! 💯

**Review by bluepnume — APPROVED** [review]

**Pull request merged**